File: 백준/Silver/1189. 컴백홈/컴백홈.cc
#include <bits/stdc++.h>
using namespace std;

int r, c, k, visited[5][5], dy[] = { 0, 0, 1, -1 }, dx[] = { 1, -1, 0, 0 }, ans;
char a[5][5];

void dfs(int y, int x, int cnt) {
	if (y == 0 && x == c - 1 && cnt == k) {
		ans++;
		return;
	}

	for (int i = 0; i < 4; i++) {
		int ny = y + dy[i];
		int nx = x + dx[i];
		if (ny >= 0 && nx >= 0 && ny < r && nx < c && a[ny][nx] != 'T' && !visited[ny][nx]) {
			visited[ny][nx] = 1;
			dfs(ny, nx, cnt + 1);
			visited[ny][nx] = 0;
		}
	}
}


int main()
{
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> r >> c >> k;
	for (int i = 0; i < r; i++) {
		for (int j = 0; j < c; j++)
			cin >> a[i][j];
	}

	visited[r-1][0] = 1;
	dfs(r-1, 0, 1);

	cout << ans;

	return 0;
}

File: 백준/Silver/1189. 컴백홈/컴백홈.java
import java.io.*;
import java.util.*;

public class Main {
	static int r, c, k, ans;
	static char[][] a;
	static boolean[][] visited;
	static int[] dy = {-1, 0, 1, 0};
	static int[] dx = {0, 1, 0, -1};

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

		StringTokenizer st = new StringTokenizer(br.readLine());
		r = Integer.parseInt(st.nextToken());
		c = Integer.parseInt(st.nextToken());
		k = Integer.parseInt(st.nextToken());
		a = new char[r][c];

		for (int i = 0; i < r; i++) {
			a[i] = br.readLine().toCharArray();
		}

		visited = new boolean[r][c];
		visited[r - 1][0] = true;
		dfs(r - 1, 0, 1);

		bw.write(ans + "");
		bw.flush();
	}

	static void dfs(int y, int x, int d) {
		if (d == k) {
			if (y == 0 && x == c - 1) {
				ans++;
			}
			return;
		}

		for (int i = 0; i < 4; i++) {
			int ny = y + dy[i];
			int nx = x + dx[i];
			if (ny < 0 || nx < 0 || ny >= r || nx >= c)
				continue;
			if (a[ny][nx] == 'T' || visited[ny][nx])
				continue;
			visited[ny][nx] = true;
			dfs(ny, nx, d + 1);
			visited[ny][nx] = false;
		}
	}
}
